examples/cmd/logtransport.go

```diff
@@ -22,15 +21,23 @@ type logTransport struct {
 func (t *logTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	t0 := time.Now()
 
-	reqBody, _ := ioutil.ReadAll(r.Body)
-	r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
-
+	// Since the underlaying Transport.RoundTrip might consume the Request, and it's also needed later to log the process,
+	// it is needed to make an snapshot of the Request.Body in advance; otherwise, its will be no longer accessible.
+	bakupReqBody := utils.BackupRequestBody(r)
 	resp, err := t.T.RoundTrip(r)
+	r.Body = bakupReqBody
+
+	// Since the outter Transport.RoundTrip calling this one might also consume the Request, and it is also needed here to log
+	// the process, it will be done over a snapshot of the Request.Body; otherwise, its will be no longer accessible by outer Transports.
+	reqBody, _ := utils.ReadRequestAndRestore(r)
+
 	if err != nil {
 		return resp, err
 	}
 
-	respBody, err := ioutil.ReadAll(resp.Body)
+	// Since the Response has been read, it is needed to restore it
+	// before reusing it; otherwise, its body will be unreadable again in other Transports.
+	respBody, err := utils.ReadResponseAndRestore(resp)
```

Comment:
Logger shouldn't care about what other transports do. It gets the response and it logs it. If for some reason the response is empty it should log empty.

Reply:
The logger should log the response **without ruining it**; that's because it needs to:
1. read it,
1. and then restore it.

Reply:
yes. as it is done in the original implementation `resp.Body = ioutil.NopCloser(bytes.NewBuffer(respBody))`

Reply:
Exactly!
that logic is inside of `utils.ReadResponseAndRestore(resp)`, so you get the content of the response without modifying it, so without needing to restore it some lines below (converting the process of read and restore as a kind of transaction), no matter what's happening in the middle.

Reply:
okay. The comment has confused me here, the code is ok.
`Since the Response has been read` for me reads as `it has been read before this comment`.
I would propose to change the comment to:
```
// Read the body and preserve it in the response
```
or remove the comment completely as the next line kind of self-explaining.

Reply:
Totally agree, no comment is better than a redundant or misleading comment.
Removed by 24ef3bf